src/cargo/core/compiler/build_runner/compilation_files.rs

```diff
@@ -931,6 +931,10 @@ fn use_pkg_dir(bcx: &BuildContext<'_, '_>, unit: &Unit) -> bool {
         // Doc tests do not have metadata.
         return false;
     }
+    if bcx.gctx.cli_unstable().build_dir_new_layout {
+        // These always use metadata.
+        return true;
+    }
```

Comment:
There is no test showing the behavior change. Should we add some no-metadata test cases for both old and new layout, so at least new layout will show it start having metadata